plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+      deriving (Show)
+
+-- | A context is a stack of frames.
+type EvaluationContext = [Frame]
+
+-- | A local version of the EvaluationResult type using closures instead of values
+data LMachineResult
+    = Success Closure Heap  -- We need both the environment and the heap to see what the value "really" is.
+    | Failure
+
+emptyEnvironment :: Environment
+emptyEnvironment = Environment IntMap.empty
+
+-- | Extend an environment with a new binding
+updateEnvironment :: Int -> HeapLoc -> Environment -> Environment
+updateEnvironment index cl (Environment m) = Environment (IntMap.insert index cl m)
+
+-- | Look up a heap location in an environment.
+lookupHeapLoc :: Name () -> Environment -> HeapLoc
+lookupHeapLoc name (Environment env) =
+    case IntMap.lookup (unUnique $ nameUnique name) env of
+      Nothing  -> throw $ LMachineStringException ("Name " ++ show (nameString name) ++ " missing from environment")
+      Just loc -> loc
+
+emptyHeap :: Heap
+emptyHeap = Heap IntMap.empty 0
+
+-- | Insert a new entry into a heap. The heap grows monotonically with no garbage collection or reuse of heap slots.
+insertInHeap :: Closure -> Heap -> (HeapLoc, Heap)
+insertInHeap cl (Heap h top) =
+    let top' = top + 1
+    in (top', Heap (IntMap.insert top' (Unevaluated cl) h) top')
+
+-- | Replace the heap entry at a given location.
+-- The old entry presumably persits in the old heap.
+updateHeap :: HeapLoc -> HeapEntry -> Heap -> Heap
+updateHeap l cl (Heap h top) =
+    Heap (IntMap.insert l cl h) top
+    -- insert seems to be faster than adjust (18s vs 28 s for fib 32) and also uses less memory.
+
+lookupHeap :: HeapLoc -> Heap -> HeapEntry
+lookupHeap l (Heap h _) =
+    case IntMap.lookup l h of
+      Just e  -> e
+      Nothing -> throw $ LMachineStringException ("Missing heap location in lookupHeap: " ++ show l)  -- This should never happen
+
+
+-- | The basic computation step of the L machine.  Search down the AST looking for a value, saving surrounding contexts on the stack.
+computeL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+computeL ctx heap cl@(Closure term env) =
+    case term of
+      TyInst _ fun ty       -> computeL (FrameTyInstArg ty : ctx)                 heap (Closure fun env)
+      Apply _ fun arg       -> computeL (FrameAppArg (Closure arg env) : ctx)     heap (Closure fun env)
+      Wrap ann tyn ty term' -> computeL (FrameWrap ann tyn ty : ctx)              heap (Closure term' env)
+      Unwrap _ term'        -> computeL (FrameUnwrap : ctx)                       heap (Closure term' env)
+      TyAbs{}               -> returnL  ctx heap cl
+      LamAbs{}              -> returnL  ctx heap cl
+      Builtin{}             -> returnL  ctx heap cl
+      Constant{}            -> returnL  ctx heap cl
+      Error{}               -> Failure
+      Var _ name            -> let l = lookupHeapLoc name env
```

Comment:
Very nicely structured, perhaps I should change the CEK machine to also use `case`.